engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -0,0 +1,4415 @@
+      //
+      List<StepMeta> nextSteps = transMeta.findNextSteps( stepMeta );
+      if ( nextSteps.size() > 0 ) {
+        nextStepPartitioningMeta = nextSteps.get( 0 ).getStepPartitioningMeta();
+      }
+
+      // TODO: throw exception if we're not partitioning yet.
+      // For now it throws a NP Exception.
+    }
+
+    int partitionNr;
+    try {
+      partitionNr = nextStepPartitioningMeta.getPartition( rowMeta, row );
+    } catch ( KettleException e ) {
+      throw new KettleStepException(
+        "Unable to convert a value to integer while calculating the partition number", e );
+    }
+
+    RowSet selectedRowSet = null;
+
+    if ( clusteredPartitioningFirst ) {
+      clusteredPartitioningFirst = false;
+
+      // We are only running remotely if both the distribution is there AND if the distribution is actually contains
+      // something.
+      //
+      clusteredPartitioning =
+        transMeta.getSlaveStepCopyPartitionDistribution() != null
+          && !transMeta.getSlaveStepCopyPartitionDistribution().getDistribution().isEmpty();
+    }
+
+    // OK, we have a SlaveStepCopyPartitionDistribution in the transformation...
+    // We want to pre-calculate what rowset we're sending data to for which partition...
+    // It is only valid in clustering / partitioning situations.
+    // When doing a local partitioning, it is much simpler.
+    //
+    if ( clusteredPartitioning ) {
+
+      // This next block is only performed once for speed...
+      //
+      if ( partitionNrRowSetList == null ) {
+        partitionNrRowSetList = new RowSet[ outputRowSets.size() ];
+
+        // The distribution is calculated during transformation split
+        // The slave-step-copy distribution is passed onto the slave transformation
+        //
+        SlaveStepCopyPartitionDistribution distribution = transMeta.getSlaveStepCopyPartitionDistribution();
+
+        String nextPartitionSchemaName =
+          TransSplitter.createPartitionSchemaNameFromTarget( nextStepPartitioningMeta
+            .getPartitionSchema().getName() );
+
+        for ( RowSet outputRowSet : outputRowSets ) {
+          try {
+            // Look at the pre-determined distribution, decided at "transformation split" time.
+            //
+            int partNr =
+              distribution.getPartition(
+                outputRowSet.getRemoteSlaveServerName(), nextPartitionSchemaName, outputRowSet
+                  .getDestinationStepCopy() );
+
+            if ( partNr < 0 ) {
+              throw new KettleStepException( "Unable to find partition using rowset data, slave="
+                + outputRowSet.getRemoteSlaveServerName() + ", partition schema="
+                + nextStepPartitioningMeta.getPartitionSchema().getName() + ", copy="
+                + outputRowSet.getDestinationStepCopy() );
+            }
+            partitionNrRowSetList[ partNr ] = outputRowSet;
+          } catch ( NullPointerException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)